sdktests/server_side_eval_bucketing.go

```diff
@@ -0,0 +1,344 @@
+			salt:                "saltyA",
+			contextValue:        33333,
+			expectedBucketValue: 0.54771423,
+		},
+		{
+			flagOrSegmentKey:    "hashKey",
+			salt:                "saltyA",
+			contextValue:        99999,
+			expectedBucketValue: 0.7309658,
+		},
+		// Non-integer numeric values, and any value types other than string and number, are not allowed
+		// and cause the bucket value to be zero.
+		{
+			flagOrSegmentKey:    "hashKey",
+			salt:                "saltyA",
+			contextValue:        1.5,
+			expectedBucketValue: 0,
+		},
+		{
+			flagOrSegmentKey:    "hashKey",
+			salt:                "saltyA",
+			contextValue:        true,
+			expectedBucketValue: 0,
+		},
+		{
+			flagOrSegmentKey:    "hashKey",
+			salt:                "saltyA",
+			contextValue:        []interface{}{"x"},
+			expectedBucketValue: 0,
+		},
+		{
+			flagOrSegmentKey:    "hashKey",
+			salt:                "saltyA",
+			contextValue:        map[string]interface{}{"x": "y"},
+			expectedBucketValue: 0,
+		},
+	}
+}
+
+func makeBucketingTestParamsForExperiments() []bucketingTestParams {
+	ret := makeBucketingTestParams()
+	ret = append(ret, []bucketingTestParams{
+		{
+			flagOrSegmentKey:    "hashKey",
+			salt:                "saltyA",
+			contextValue:        "userKeyA",
+			expectedBucketValue: 0.42157587,
+		},
+		{
+			flagOrSegmentKey:    "hashKey",
+			salt:                "saltyA",
+			contextValue:        "userKeyB",
+			expectedBucketValue: 0.6708485,
+		},
+		{
+			flagOrSegmentKey:    "hashKey",
+			salt:                "saltyA",
+			contextValue:        "userKeyC",
+			expectedBucketValue: 0.10343106,
+		},
+		{
+			flagOrSegmentKey:    "hashKey",
+			salt:                "saltyA",
+			contextValue:        "userKeyA",
+			seed:                ldvalue.NewOptionalInt(61),
+			expectedBucketValue: 0.09801207,
+		},
+		{
+			flagOrSegmentKey:    "hashKey",
+			salt:                "saltyA",
+			contextValue:        "userKeyB",
+			seed:                ldvalue.NewOptionalInt(61),
+			expectedBucketValue: 0.14483777,
+		},
+		{
+			flagOrSegmentKey:    "hashKey",
+			salt:                "saltyA",
+			contextValue:        "userKeyC",
+			seed:                ldvalue.NewOptionalInt(61),
+			expectedBucketValue: 0.9242641,
+		},
+	}...)
+	return ret
+}
+
+func RunServerSideEvalBucketingTests(t *ldtest.T) {
+	unwantedVar, expectedFallthroughVar, expectedRuleVar, expectedSegmentVar := 0, 1, 2, 3
+	expectedFallthroughValue := ldvalue.String("expected-value-for-fallthrough")
+	expectedRuleValue := ldvalue.String("expected-value-for-rule")
+	expectedSegmentValue := ldvalue.String("expected-value-for-segment-match")
+	unwantedValue, defaultValue := ldvalue.String("wrong"), ldvalue.String("default")
+	matchRuleAttr := "matchrule"
+
+	makeRolloutVariationsToMatch := func(expectedBucketValue float32, desiredVariation int) []ldmodel.WeightedVariation {
+		targetBucketSize := 10 // arbitrary small number (0.01% of 100000) to allow for a tiny bit of rounding error
```

Comment:
This is pretty arbitrary and I'm not sure how much variance we want to tolerate. I mean, theoretically _none_, but since all of the SDKs currently converting integer weights to fractions by dividing by a power of 10, which can't ever be exact in binary floating point, I'm not sure how feasible that would be.